### PR TITLE
feat(text-inputs): add new validations option to text input components FE-4285

### PR DESCRIPTION
**Pull request by samtjo**
### Proposed behaviour

This PR adds (opt in) support for the new validation designs in text input components. The steps to add this were: 

- Add a new internal component (`ValidationMessage`) to be reused in other components.
- Add a `validationRedesignOptIn` prop to `CarbonProvider`. This is `false` by default.
- Update text input components to align with the new design when `validationRedesignOptIn` is set to `true`.

The following components now support the new validation designs: 

- Textbox
- Date
- DateRange
- Decimal
- GroupedCharacter
- Number
- Textarea
- Select
- FilterableSelect
- MultiSelect
- NumeralDate

The option to have the new validation designs on `Fieldset` has been completely disabled for now.

Also increase the spacing between the two inputs on `DateRange` to `24px` from `12px`.

Example of new design: 

image

image

### Current behaviour

Currently we don't support the new validation design in any of our components.

### Checklist

- [x] Commits follow our style guide
- [x] Related issues linked in commit messages if required
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [ ] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [x] Typescript `d.ts` file added or updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Testing instructions

See the new Chromatic snapshots for new validation support in each of the components listed above.

To test manually: 

http://localhost:9001/?path=/story/date-range--validations-string-new-design
http://localhost:9001/?path=/story/date-input--validations-string-new-design
http://localhost:9001/?path=/story/decimal-input--validations-string-new-design
http://localhost:9001/?path=/story/groupedcharacter--validations-string-new-design
http://localhost:9001/?path=/story/number-input--validations-string-new-design
http://localhost:9001/?path=/story/numeral-date--validations-string-new-design
http://localhost:9001/?path=/story/select--validations-string-new-design
http://localhost:9001/?path=/story/select-filterable--validations-string-new-design
http://localhost:9001/?path=/story/select-multiselect--validations-string-new-design
http://localhost:9001/?path=/story/textarea--validations-string-new-design
http://localhost:9001/?path=/story/textbox--validations-string-new-design

### CODE REVIEW

File: src/components/date/date-test.stories.mdx [hunk 1]
```diff
@@ -31,14 +32,40 @@ export const DateStory = (args) => {
...
+        action("onBlur")(ev.target.value);
+      }}
       onKeyDown={(ev) => action("onKeyDown")(ev.target.value)}
       onClick={(ev) => action("onClick")(ev.target.value)}
       {...getCommonTextboxArgsWithSpecialCaracters(args)}
     />
   );
 };
 
+export const NewValidationStory = (args) => {
```
Comment (Parsium) [inline src/components/date/date-test.stories.mdx:45]: In Storybook, the validation message doesn't show by default nor there is a way to trigger it, is this intentional @samtjo? Would QA need this?

File: src/components/decimal/decimal-test.stories.mdx [hunk 2]
```diff
@@ -99,6 +100,25 @@ export const PostStory = ({ action: actionArg, ...args }) => {
   );
 };
 
+export const NewValidationStory = (args) => {
```
Comment (Parsium) [inline src/components/decimal/decimal-test.stories.mdx:103]: Similarly with this story too

File: src/components/date/date-test.stories.mdx (hunk 1 above)
Comment (samtjo) [reply src/components/date/date-test.stories.mdx:45]: Just use the error or warning control to type a message in:

image

File: src/components/decimal/decimal-test.stories.mdx (hunk 2 above)
Comment (samtjo) [reply src/components/decimal/decimal-test.stories.mdx:103]: Same as above 👍🏻

File: src/components/select/filterable-select/filterable-select.stories.mdx [hunk 3]
```diff
@@ -637,6 +638,76 @@ If there is no `id` prop specified on an object, then the exact objects will be
   </Story>
 </Canvas>
 
+#### New designs validation
+
+<Preview>
```
Comment (Parsium) [inline src/components/select/filterable-select/filterable-select.stories.mdx:643]: There is a missing import here for `Preview`

File: src/components/select/multi-select/multi-select.stories.mdx [hunk 4]
```diff
@@ -492,6 +493,78 @@ See <LinkTo kind="Documentation/Colors" name="page" story="page">Colors</LinkTo>
   </Story>
 </Canvas>
 
+#### New designs validation
+
+<Preview>
```
Comment (Parsium) [inline src/components/select/multi-select/multi-select.stories.mdx:498]: Same here :)

File: src/components/select/simple-select/simple-select.stories.mdx [hunk 5]
```diff
@@ -652,6 +653,76 @@ The inline label can change to be top aligned at a breakpoint. Enable this by pa
   </Story>
 </Canvas>
 
+#### New designs validation
+
+<Preview>
```
Comment (Parsium) [inline src/components/select/simple-select/simple-select.stories.mdx:658]: Same here too, storybook also crashes when clicking to open select list?

Comment (samtjo) [reply src/components/select/simple-select/simple-select.stories.mdx:658]: That's an issue on master 👍🏻 https://github.com/Sage/carbon/issues/4878

File: src/components/textbox/textbox.stories.mdx [hunk 6]
```diff
@@ -116,19 +125,43 @@ import Textbox from "carbon-react/lib/components/textbox";
   </Story>
 </Canvas>
 
+### Prefix
+
+<Preview>
```
Comment (Parsium) [inline src/components/textbox/textbox.stories.mdx:130]: Imports missing here too :)

File: src/components/select/filterable-select/filterable-select.stories.mdx (hunk 3 above)
Comment (samtjo) [reply src/components/select/filterable-select/filterable-select.stories.mdx:643]: Yeah these should have been `Canvas` not `Preview` - copy and paste error!

File: src/components/textarea/textarea.spec.js [hunk 7]
```diff
@@ -413,6 +416,74 @@ describe("componentWillUnmount", () => {
...
+            marginBottom: "8px",
+            color: "var(--colorsUtilityYin055)",
+          },
+          wrapper.find(StyledHintText)
+        );
+      });
+    });
+
+    describe("new validation design", () => {
+      it("error message is visible when the prop is passed", () => {
```
Comment (Parsium) [inline src/components/textarea/textarea.spec.js:474]: Would it be worth tweaking these two tests slightly - such that it checks the message and type (error or warning) is correct too? In my opinion these could be more explicit :)

Comment (samtjo) [reply src/components/textarea/textarea.spec.js:474]: I don't think so no as Chromatic will capture the difference between error and warning for us, good idea though 👍🏻

File: src/components/textbox/textbox.spec.js [hunk 8]
```diff
@@ -333,4 +324,72 @@ describe("Textbox", () => {
...
+            marginBottom: "8px",
+            color: "var(--colorsUtilityYin055)",
+          },
+          wrapper.find(StyledHintText)
+        );
+      });
+    });
+
+    describe("new validation design", () => {
+      it("error message is visible when the prop is passed", () => {
```
Comment (Parsium) [inline src/components/textbox/textbox.spec.js:382]: Similarly here, as mentioned in `TextArea` :)

**Review by Parsium — CHANGES_REQUESTED** [review]
Looks great @samtjo, just need to sort out the `Preview` imports in the MDX files 👍🏼

File: src/components/numeral-date/numeral-date.component.js [hunk 9]
```diff
@@ -193,8 +199,18 @@ const NumeralDate = ({
           fieldHelp={fieldHelp}
           adaptiveLabelBreakpoint={adaptiveLabelBreakpoint}
           isRequired={required}
+          validationRedesignOptIn={validationRedesignOptIn}
           {...filterStyledSystemMarginProps(rest)}
         >
+          {validationRedesignOptIn && labelHelp && (
+            <StyledHintText>{labelHelp}</StyledHintText>
```
Comment (mkrds) [inline src/components/numeral-date/numeral-date.component.js:206]: `StyledHintText` and `ValidationMessage` components are meant to be placed in `FormField` in the future when we also redesign the other inputs?

File: src/components/textbox/textbox.stories.mdx [hunk 10]
```diff
@@ -291,6 +324,42 @@ It is possible to use the `tooltipPosition` to override the default placement of
   </Story>
 </Canvas>
 
+#### New designs validation
+
+<Preview>
```
[the remark this answers is not in the document]
Comment (samtjo) [reply src/components/textbox/textbox.stories.mdx:329]: Copy and paste error, forgot to change them

File: src/__internal__/validation-message/validation-message.component.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
+import React from "react";
```
[the remark this answers is not in the document]
Comment (samtjo) [reply src/__internal__/validation-message/validation-message.component.js:1]: True, but I inherited this commit from Rafal before we added TypeScript support. It won't take me long to change it actually so I will 👍🏻

File: src/components/numeral-date/numeral-date.component.js (hunk 9 above)
Comment (samtjo) [reply src/components/numeral-date/numeral-date.component.js:206]: Yeah that would make sense to me 👍🏻

**Review by mkrds — APPROVED** [review]

**Review by Parsium — APPROVED** [review]

**Review by Parsium — APPROVED** [review]

**Review by mkrds — DISMISSED** [review]

**Review by sianford — APPROVED** [review]

**Review by Parsium — APPROVED** [review]

**Review by mkrds — APPROVED** [review]

**Pull request merged**